### PR TITLE
add esp32 build to CI

### PR DESCRIPTION
**Pull request by davidbuzz**

### CODE REVIEW

**Review by tridge — APPROVED** [review]

**Comment by davidbuzz** [comment]
my force-push adds a newer version of cmake,  makes it a bit less verbose on the console, captures the binaries for plane and copter after each build, and uploads them as artifacts.     It also fixes a small remaining issue in the bootloader build overflowing  its allocated partition space with some compiler/s.

**Review by peterbarker — COMMENTED** [review]
@davidbuzz this really ought to be two commits now.

... the current commit message should probably be `.github: ...`

**Comment by davidbuzz** [comment]
...does a little dance.. ci is passed....

**Review by peterbarker — APPROVED** [review]

File: .github/workflows/esp32_build.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          sudo apt-get install git wget libncurses-dev flex bison gperf python3 python3-pip python3-setuptools python3-serial python3-cryptography python3-future python3-pyparsing python3-pyelftools cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0
+          sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 10
+
+          rm -rf /usr/local/bin/cmake
+
+          sudo apt remove --purge --auto-remove cmake
+          sudo apt update && \
+          sudo apt install -y software-properties-common lsb-release && \
+          sudo apt clean all
+          wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | gpg --dearmor - | sudo tee /etc/apt/trusted.gpg.d/kitware.gpg >/dev/null
```
Comment (khancyr) [inline .github/workflows/esp32_build.yml:35]: why using this Cmake version ? 
thus it doesn't match kiware instructions : https://apt.kitware.com/ ?

File: .github/workflows/esp32_build.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        gcc: [10]
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+      - name: Install Prerequisites
+        shell: bash
+        run: |
+          sudo apt-get install git wget libncurses-dev flex bison gperf python3 python3-pip python3-setuptools python3-serial python3-cryptography python3-future python3-pyparsing python3-pyelftools cmake ninja-build ccache libffi-dev libssl-dev dfu-util libusb-1.0-0
```
Comment (khancyr) [inline .github/workflows/esp32_build.yml:26]: if you removed if after, just don't installl cmake there

File: .github/workflows/esp32_build.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        env:
+          CI_BUILD_TARGET: ${{matrix.config}}
+        shell: bash
+        run: |
+          source ~/.bash_profile
+          PATH="/github/home/.local/bin:$PATH"
+          echo $PATH
+          
+          cd modules/esp_idf
+          ./install.sh
```
Comment (khancyr) [inline .github/workflows/esp32_build.yml:62]: install again ?

**Review by khancyr — COMMENTED** [review]
worth doing a docker image for this on a second step!

File: .github/workflows/esp32_build.yml (hunk 1 above)
Comment (davidbuzz) [reply .github/workflows/esp32_build.yml:35]: cmake-upgrade taken verbatim from the top answer here: https://askubuntu.com/questions/355565/how-do-i-install-the-latest-version-of-cmake-from-the-command-line

File: .github/workflows/esp32_build.yml (hunk 2 above)
Comment (davidbuzz) [reply .github/workflows/esp32_build.yml:26]: the prerequisites list is taken verbatim from the ESP IDF instructions here: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/linux-macos-setup.html#get-started-prerequisites

File: .github/workflows/esp32_build.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | gpg --dearmor - | sudo tee /etc/apt/trusted.gpg.d/kitware.gpg >/dev/null
+          sudo apt-add-repository "deb https://apt.kitware.com/ubuntu/ $(lsb_release -cs) main"
+          sudo apt update
+          sudo apt install cmake
+
+
+          git submodule update --init --recursive
+          ./Tools/scripts/esp32_get_idf.sh
+
+          sudo ln -s /usr/bin/ninja /usr/bin/ninja-build
```
[the remark this answers is not in the document]
Comment (davidbuzz) [reply .github/workflows/esp32_build.yml:44]: nope, its not.

File: .github/workflows/esp32_build.yml (hunk 3 above)
Comment (davidbuzz) [reply .github/workflows/esp32_build.yml:62]: ./install.sh and 'source export.sh' together set up a bunch of PATHS and env vars [such as a python virtual env and more ] that don't persist between the different execution blocks in the yml file, so the first time its run[in the 'prerequisities' block] , it downloads everything, and the second time its run, its quicker as its already done.

File: .github/workflows/esp32_build.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (khancyr) [reply .github/workflows/esp32_build.yml:44]: https://packages.ubuntu.com/focal/amd64/ninja-build/filelist

**Pull request merged**